src/HaskellWorks/Data/Aeson.hs

```diff
@@ -60,3 +62,37 @@ readJson t s = case readMaybe s of
 -- @
 objectEndo :: [JsonEndo Pair] -> Value
 objectEndo es = object $ unJsonEndo (mconcat es) []
+
+-- | Generate key values from a value of a type.  This can be used
+-- in conjunction with 'WithJsonKeyValues' to define a 'ToJSON' instance
+-- without having to implement both 'toJSON' and 'toEncoding', thereby
+-- reducing boilerplate.
+--
+-- For example:
+--
+-- @
+-- instance ToJsonEncoding MyType where
+--   toJsonEncoding sv =
+--     [ "my_field" .!= sv ^. #myField
+--     ]
+-- @
+class ToJsonKeyValues a where
+  toJsonKeyValues :: (KeyValue kv, Monoid kv) => a -> [kv]
```

Comment:
What about things that aren't always objects in JSON? Often when dealing with APIs you'll have to deal with a union type like

```haskell
data T = Foo | Bar { f1 :: A, f2 :: b, f3 :: C }

instance toJSON T where
  toEncoding Foo = toEncoding @ByteString "foo"
  toEncoding (Bar f1 f2 f3) = pairs $ "f1" .= f1 <> "f2" .= f2 <> "f3" .= f3
  toJSON Foo = toJSON $ String "foo"
  toJSON (Bar f1 f2 f3) = object ["f1" .= f1, "f2" .= f2, "f3" .= f3]
```

Reply:
In that case you could write out the `ToJSON` instance as you have.  One could also have multiple `deriving via` `newtype`s for various cases.